### ISSUE TITLE
@emotion/react ThemeProvider does not work with @emotion/styled

### ISSUE THREAD

**Issue by fabian-hiller** [storybookjs/storybook#10231]
**Describe the bug**
I use @emotion/react ThemeProvier to access the theme prop when styling a component. When I open my app with "npm start" everything works smoothly. With "npm run storybook", however, the theme object for my styled emotion components is empty.

**To Reproduce**
1. npx create-react-app my-app
2. npx -p @storybook/cli sb init
3. npm install @emotion/react@11.0.0-next.12 @emotion/styled@11.0.0-next.12
4. Creat Theme and Button component with logs for the theme prop
5. Creat Button.stories.js
6. Add Theme and Button component to App.js
7. Add Theme component (with addDecorator) as decorator to .storybook/preview.js
8. Run "npm start" and "npm run storybook"

**Expected behavior**
The pre-defined theme object is displayed in the console, so that you can access it when styling.

**Screenshots**
![image](https://user-images.githubusercontent.com/35291865/77758468-4fa3cf80-7033-11ea-8cb9-deb17f2f9e1b.png)

**Code snippets**
GitHub repo: https://github.com/fabian-hiller/a7hs5hsk2

**System:**
Environment Info:
  System:
    OS: macOS 10.15.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 12.13.1 - /usr/local/bin/node
    npm: 6.14.4 - ~/.npm-global/bin/npm
  Browsers:
    Chrome: 80.0.3987.149
    Firefox: 74.0
    Safari: 13.0.5
  npmPackages:
    @storybook/addon-actions: ^5.3.17 => 5.3.17 
    @storybook/addon-links: ^5.3.17 => 5.3.17 
    @storybook/addons: ^5.3.17 => 5.3.17 
    @storybook/preset-create-react-app: ^2.1.1 => 2.1.1 
    @storybook/react: ^5.3.17 => 5.3.17

**Comment by shilman** [comment]
We are also using emotion 10.0.x in Storybook. I'm guessing the two are not playing nicely together. cc @ndelangen

**Comment by fabian-hiller** [comment]
Hello @shilman , thank you very much for the fast processing. I am currently in the process of moving our design system to Storybook. Due to this problem, a part of our frontend development is currently at a standstill. If it helps you, I will support you in solving this problem.

**Comment by fabian-hiller** [comment]
Is there any news yet? @shamin @ndelangen

**Comment by shilman** [comment]
I don't know what's happening here. Hopefully @ndelangen  does.

**Comment by ndelangen** [comment]
Are you wrapping your stories with a decorator providing the theme context?

**Comment by fabian-hiller** [comment]
> Are you wrapping your stories with a decorator providing the theme context?

Yes. As described in step 7 (steps to reproduce), I have done that. Because the useTheme hook works, as you can see in the photo, the theme context is available.

I created a repository where you can look at the code or clone it for testing.
https://github.com/fabian-hiller/a7hs5hsk2

**Comment by tajo** [comment]
Might be related to https://github.com/storybookjs/storybook/issues/10296

It seems that context provider doesn't work on the top level in stories.

**Comment by fabian-hiller** [comment]
> It seems that context provider doesn't work on the top level in stories.

In my case it just does not work with the styled components. In the other cases, as you can see in the screenshot, the context works.

But maybe the reason for the problem is the same for both of us.

**Comment by ndelangen** [comment]
I spoke to one of the emotion maintainers yesterday, and this is essentially due to both storybook & your app/component using emotion & theming.

Emotion theming solution uses React context, meaning is somewhat global.

The proper solution is to not use emotion context but our own.

Here's the suggested solution from https://github.com/Andarist:

```js
styled.div(props => ({
  const theme = useMyTheme();
  return {
    color: theme.color,
  };
})}
```

I'm not mentioning this because I think YOU should do this, but actually, this is something Storybook should do, so we would never conflict with someone using emotion.

We think emotion is the best, and we certainly don't want to stand in the way of anyone using it + storybook.

@fabian-hiller I have some pretty strong ideas on how to make this work from storybook, would you be interested in working with me to solve this in storybook?

**Comment by fabian-hiller** [comment]
> @fabian-hiller I have some pretty strong ideas on how to make this work from storybook, would you be interested in working with me to solve this in storybook?

 Yeah, I'd be honored. How much effort do you estimate?

**Comment by ndelangen** [comment]
We need 
- to expose a custom context for theming that derives the theme from the manager-context state
- replace the ThemeProvider from emotion with this custom ContextProvider
- change @storybook/theming so `styled` & `useTheme` auto-inject them theme into any functions

part 3 is going to be the most challenging, particularly to make all the typings correct.

If you'd want to tackle this together in a pair programming session, let's schedule something:
https://calendly.com/ndelangen/storybook

**Comment by fabian-hiller** [comment]
> If you'd want to tackle this together in a pair programming session, let's schedule something:
> https://calendly.com/ndelangen/storybook

All right. See you later.

**Comment by connor-baer** [comment]
Woohoo! It's great to see this being worked on. We face the same issue at [sumup-oss/circuit-ui](https://github.com/sumup-oss/circuit-ui) — if there's anything that I can do to help here, please let me know.

**Comment by fabian-hiller** [comment]
Thanks for your message @connor-baer.

> If there's anything that I can do to help here, please let me know.

Could you find a solution? Where can I find the code for it in your repository?

**Comment by connor-baer** [comment]
@fabian-hiller I searched the Emotion and Storybook issues extensively and there doesn't seem to be a solution that can be implemented on the user's side. 

I came across the same suggestion that @ndelangen relayed from [@Andarist](https://github.com/Andarist): frameworks should implement and use their own theme context. The steps that @ndelangen outlined are much more straightforward than what I had in mind. Did you two already get started on those?

I looked a bit into [@storybook/theming](https://github.com/storybookjs/storybook/blob/next/lib/theming/src/index.ts) and how we could auto-inject the custom theme context into `styled` & `useTheme` (step 3). Looks like we need to create our own `styled` instance from [@emotion/styled-base](https://github.com/emotion-js/emotion/tree/master/packages/styled-base), something like this (pseudo code):

```tsx
import React from 'react';
import emotionStyled from '@emotion/styled-base'

const useTheme = () => React.useContext(StorybookThemeContext);

const styled = (...args) => {
  const Styled = emotionStyled(...args);
  const theme = useTheme();
  return React.forwardRef((props) => <Styled {...props} theme={theme} />;
}
```

Then we just need to apply the [same magic](https://github.com/emotion-js/emotion/blob/master/packages/styled/src/index.js) that @emotion/styled does to make the `styled.<tag>` syntax work.

**Comment by fabian-hiller** [comment]
@connor-baer I had an approach similar to yours, but I haven't been able to make it work yet. I will test your code the days and if it works, I can create a PR.

**Comment by lucascurti** [comment]
Doing something like this would be as a workaround would be a bad idea?

```jsx
import styled from "@emotion/styled";
import { theme } from "custom-theme";

export const Button = styled.button`
  background-color: ${theme.colors.green};
`;

```

**Comment by connor-baer** [comment]
> Doing something like this as a workaround would be a bad idea?

This would require updating every component inside Storybook or your own codebase which would be a  _huge_ amount of work. It also makes it impossible to dynamically update the theme, e.g. from light to dark.

> Another thing that seems to work is keep using `ThemeProvider` from `emotion-theming` until this is fixed for `@emotion/react`

I assume you're talking about using Emotion 11 in the component library? Does `emotion-theming` work with Emotion 11? I figured it wouldn't since `@emotion/styled` imports the theme context from `@emotion/react`, doesn't it?

**Comment by lucascurti** [comment]
Good point on #1.

Regarding #2, Yes. I;m using Emotion 11 and seems that using `ThemeProvider` from `emotion-theming` works. This is how I ended up doing it.

**Comment by connor-baer** [comment]
Interesting, I'll give that a shot. Thanks for the tip!

**Comment by grncdr** [comment]
Given that the above PR is not yet merged, I think this issue should be re-opened.

**Comment by RuudBurger** [comment]
Just want to mention this is still an issue. The theme works in components, but not directly in the story. We are using MDX for documentation, if that makes any difference.

The following code shows everything you need to know:

```mdx
import { useContext } from 'react';
import { ThemeContext } from '@emotion/core';

<Preview>
  <Story name="Storyname">
    {() => {
      console.log('directly in story', useContext(ThemeContext));
      const Test = () => {
        console.log('Inside any component', useContext(ThemeContext));
        return null;
      };
      return <Test />;
    }}
  </Story>
</Preview>
```

Then the logs look like this:
![Screen Shot 2020-10-26 at 11 37 16](https://user-images.githubusercontent.com/38136/97162508-9bf8ee80-177f-11eb-90ed-8f95b830c672.png)

**Comment by mverissimo** [comment]
After some research, I find a temporary solution used by **chackra-ui** [here](https://github.com/chakra-ui/chakra-ui/blob/a5abb6f9477d87a1cbc0c2d784e009d2bc8a8c6d/.storybook/main.js)! How emotion doesn't have breaking changes, this works fine.

I think storybook need to update to emotion 11 internally.

**Comment by MikeAliG** [comment]
Hello, any update on this?

**Comment by mavlikwowa** [comment]
I changed webpack.config in Storybook and it helped!
`const path = require("path")

const toPath = (_path) => path.join(process.cwd(), _path)

module.exports = {
  stories: ['../stories/**/*.stories.@(ts|tsx|js|jsx)'],
  addons: ['@storybook/addon-links', '@storybook/addon-essentials'],
  // https://storybook.js.org/docs/react/configure/typescript#mainjs-configuration
  typescript: {
    check: true, // type-check stories during Storybook build
  },
  webpackFinal: async (config) => {
    return {
      ...config,
      resolve: {
        ...config.resolve,
        alias: {
          ...config.resolve.alias,
          "@emotion/core": toPath("node_modules/@emotion/react"),
          // You should add this row
          "@emotion/styled": toPath("node_modules/@emotion/styled"),
          "emotion-theming": toPath("node_modules/@emotion/react"),
        },
      },
    }
  },
};
`

**Comment by MikeAliG** [comment]
Thanks for the reply @mavlikwowa but we tried this before and since we're working in yarn workspaces with storybook in subfolders, this method doesn't work for us, I think it would be nice if storybook provided an stable solution as per #13300. If we can help with anything, please let us know, we'll be happy to help.

**Comment by DominicTobias-b1** [comment]
This problem isn't related to emotion v11 it happens in v10 when you have a Storybook of components which also have a theme provider. Unfortunately unless you wrap the app in BOTH theme providers you get `{}` in the styled components for the theme object.

I have a hunch that this is causing an issue for us though since we can override styles of a styled component in storybook via passing a prop into `css={someProp}`, but when doing it in the app it doesn't take precedence.

**Comment by nrakochy** [comment]
Are there any other work-around options? 

The `emotion` alias option does not seem to be working in my yarn project as well. Further, after downgrading `theme-ui` and `emotion` to get access to a working provider, I am [running into this error](https://github.com/system-ui/theme-ui/issues/1160), which causes conflicts with `create-react-app` pragma from v4.

```js
Syntax error: pragma and pragmaFrag cannot be set when runtime is automatic
```

To say it another way, this downgrade is untenable and I am now effectively stuck. I am happy to contribute if there is any way to make the `ThemeProvider` just work. And open to any other suggestions to get unblocked. Many thanks!

**Comment by zereight** [comment]
same to me

**Comment by tihuan** [comment]
https://github.com/storybookjs/storybook/issues/10231#issuecomment-728038867 this temp solution works for me too!

**Comment by ndelangen** [comment]
That solution seems to force Emotion11 upon storybook, right?

Perhaps our inline rendering in docs could be change to use separate rendering, so the context between storybook and the story is separate too?

## Here's a concept:

We render the `DocsContainer` & `Page` here:
https://github.com/storybookjs/storybook/blob/62bc50560d6c4e8d191556f569519b6c4d406ff9/lib/core-client/src/preview/StoryRenderer.tsx#L313-L318

We render the inline story here:
https://github.com/storybookjs/storybook/blob/04acc44510fc189003e643c3262e1db46275f859/lib/components/src/blocks/Story.tsx#L41-L43

If instead of rendering directly:
```jsx
<Fragment>
  {storyFn ? createElement(storyFn) : <EmptyBlock>{MISSING_STORY(id)}</EmptyBlock>}
</Fragment>
```

We could render a placeholder:
```jsx
<div id=`render-location-story-${id}`></div>
```

And call an util function for rendering with that identifier via `useEffect`..:
```js
useEffect(() => { renderStoryAtNode(id) }, [id]);
```

`renderStoryAtNode` will lookup what it needs to based on the and perform `ReactDOM.render` it's own !

I think this solves ALL context based issues with storybook & the user's app sharing React context.

`renderStoryAtNode` could be based on an event, so we don't have to do dependency injection.
(the render function is different in every framework, so where would `renderStoryAtNode` from?)

So this is what it would really look like:
```js
useEffect(() => { emit('renderStoryAtNode', id) }, [id]);
```

---

I also wonder if this means we can render non-react inline? Given how we'd detach rendering docs from the actual story.

**Comment by ndelangen** [comment]
## potential complications:

- dynamic story source rendering ?
- render multiple instances of framework (for examples angular)
  solved in the existing inline rendering solution for example Angular (@ThibaudAV)

**Comment by ndelangen** [comment]
## potential additional upside:

- No more need for the whole `prepareForInline` business (converting non-react to react)

**Comment by HoverBaum** [comment]
Also ran into this issue today and it took some time to figure out which combination of things was causing the issue. I am building a component library upon styled-system and emotion. Inside my App, everything worked fine but in Storybook the components didn't take any values from the ThemeProvider (added via decorators in preview.js).
My workaround, for now, is to `import { ThemeProvider } from 'emotion-theming'`. As suggested by [lucascurty](https://github.com/storybookjs/storybook/issues/10231#issuecomment-613394048).

Would love to see support for ThemeProvider from @emotion/react 🙂

**Comment by badsyntax** [comment]
For me the solution was to add the emotion packages to the webpack.config.js:

```js
module.exports = {
  resolve: {
    extensions: ['.ts', '.tsx', '.js'],
    alias: {
      ...,
      '@emotion/react': '@emotion/react',
      '@emotion/styled': '@emotion/styled',
    },
  },
  ...
```

**Comment by fslc92** [comment]
I have it working like this

**.storybook/main.js**
```
const path = require("path")
const { override, babelInclude, addWebpackAlias, addWebpackResolve } = require("customize-cra")

module.exports = {
  stories: [
    "../stories/**/*.js"
  ],
  addons: [
    "@storybook/addon-links",
    {
      name: "@storybook/addon-essentials",
      options: {
        actions: false,
        controls: false,
        viewport: false,
      }
    },
    "@react-theming/storybook-addon"
  ],
  webpackFinal: async (config, env) => {  
    return Object.assign(
      config,
      override(
        /* TODO: remove once this is merged: https://github.com/storybookjs/storybook/pull/13300
         *
         * Storybook does not support emotion 11 yet.
         * Sometimes the modules get crossed. If that happens, uncomment these lines, delete/re-install modules and run storybook again.
         * IT WILL FAIL until you re-comment out these lines and restart the storybook server.
         * import addWebpackAlias from customize-cra
         *
        **/
        addWebpackAlias({
          "@emotion/core": path.resolve("node_modules/@emotion/react"),
          "@emotion/styled": path.resolve("node_modules/@emotion/styled"),
          "emotion-theming": path.resolve("node_modules/@emotion/react"),
          "@emotion/react": path.resolve("node_modules/@emotion/react"),
        }),
        addWebpackResolve({
          modules: [
            ...(config.resolve.modules || []),
            path.resolve(__dirname, "../../webapp/src")
          ] 
        }),
      )(config, env)
    )
  },
}
```

**.storybook/preview.js**
```
import { ThemeProvider } from "@emotion/react"
import { withThemes } from "@react-theming/storybook-addon"

const themes = []

export const decorators = [
  withThemes(ThemeProvider, themes),
]

export const parameters = {
  backgrounds: {
    default: "light",
    values: themes.map(({ name, colors }) => ({ name, value: colors.background }))
  },
}

```

**Comment by onpaws** [comment]
For anyone else having problems with MUIv5 + Storybook 'Docs' tab, this was my solution.

We use MUIv5 (currently in beta) which depends on Emotion 11. 
Some (all?) MUI components depend on the MUI theme object, commonly provided by a MUI [`<ThemeProvider>`](https://github.com/mui-org/material-ui/blob/e724d98eba018e55e1a684236a2037e24bcf050c/packages/material-ui-private-theming/src/ThemeProvider/ThemeProvider.js) 
e.g. `<OutlinedInput>` includes:
```javascript
borderColor: theme.palette.mode === 'light' ? 'rgba(0, 0, 0, 0.23)' : 'rgba(255, 255, 255, 0.23)'
```
Like many others here, I added a decorator in preview.js, but it wasn't enough
```jsx
export const decorators = [
  (Story) => (
    <MUIThemeProvider theme={muiTheme}>
      <CssBaseline />
      <Story />
    </MUIThemeProvider>
  ),
];
```
The problem: as we started trialling Storybook v6.3, while Storybook's 'Canvas' tab was working OK, something about the 'Docs' tab is different b/c components whose styling depends on values from the MUI theme (e.g. dark/light relies on`theme.palette.mode`) were crashing with `Uncaught TypeError: Cannot read properties of undefined (reading 'mode')`

Inspecting via React DevTools, it looks like Storybook wraps things using its own `ThemeProvider`, which if I understand correctly, seems to 'win' (overwrite?) vs MUI's `ThemeProvider`. 

I started with @lucascurti [tip](https://github.com/storybookjs/storybook/issues/10231#issuecomment-613394048) (thanks!),  but _also_ wrapped that with MUI's theme provider.

preview.js
```react
import { ThemeProvider as MUIThemeProvider } from "@material-ui/core";
import { ThemeProvider } from "emotion-theming";

export const decorators = [
  (Story) => (
    <MUIThemeProvider theme={muiTheme}>
      <ThemeProvider theme={muiTheme}>
        <CssBaseline />
        <Story />
      </ThemeProvider>
    </MUIThemeProvider>
  ),
];

```
I'm pretty sure this is mixing both Emotion 10 and 11, which seems not great and hope we don't have new issues down the road.
Regardless, for today at least, I'm glad Docs tab works again.

It seems preferable to have just one ThemeProvider, for this reason I'd also love to Emotion 11 support land, which I believe would make my solution no longer needed.

**Comment by HosseinAgha** [comment]
We have the exact same issue. @cabello.   
[The @mavlikwowa's solution](https://github.com/storybookjs/storybook/issues/10231#issuecomment-751239488) only makes theme available in the emotion styled-components but it looks like `useTheme` is not working.

**Comment by totszwai** [comment]
I tried using the solution from @onpaws

```
import React from 'react';
import { ThemeProvider as MuiThemeProvider, CssBaseline } from '@mui/material';
import { ThemeProvider } from 'styled-components';
import { GlobalStyle } from './index';

...

<StyledEngineProvider injectFirst>
    <MuiThemeProvider theme={props.theme}>
      <ThemeProvider theme={props.theme}>
        <GlobalStyle />
        <CssBaseline />
        {props.children}
      </ThemeProvider>
    </MuiThemeProvider>
</StyledEngineProvider>
```

But the typography setting configured for MUI is not passing down... If I run my project directly with `react-scripts` everything is working. Anyone here know how to get it to work with storybook? :(

**Comment by rrsai** [comment]
Storybook has provided a feature-flag for storybook users that works around this issue for now, see the other issue: https://github.com/storybookjs/storybook/issues/13145  For those who spend time looking for a fix for this after me.

Or to skip to the docs:

> After much debugging of the clash between versions of emotion used in storybook and in app, and for those who come after me, I found that storybook has now provided a feature flag to avoid clashing with later versions of emotion: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#emotion11-quasi-compatibility
> I tested it and it works well to prevent emotion version-out-of-sync problems in storybook, and is a 3 line change.

**Comment by shilman** [comment]
@rrsai thanks so much for spreading the word!!! 🙏 🙏 🙏

**Comment by azizoid** [comment]
I ran to that problem recently, so everything that need to be upgraded need to be upgraded
